elastic/beats
# Support serverless

This is a rather large PR that updates beats to provide DSL support and setup on serverless. Some of the parts are:

- Update the config templates to add a `setup.dsl.*` section
- Renaming a considerable number of files in libbeat/idxmgmt, either to remote "ILM" or for the sake of clarity
- Refactor of the lifecycle managers and index manages to support serverless awareness as part of the setup process
- Major refactor of the setup client handlers, moving large parts of the index management process to the client handlers, to enable simpler switching between ILM/DSL
- Changes to the index template process to enable the DSL setup
- Additional tests

As an added note: the index management code is pretty complex and labyrinthine, and while I tried to simplify some things, it's still (needlessly) complex. I'm fairly certain that it would be possible to significantly simplify a lot of this code, but due to the tight deadlines of serverless support, it probably won't happen as part of this PR.

## Checklist



## How to test this PR locally

When I started working on this, my hope was that I could just glue on DSL support without touching the existing ILM code, but that wasn't the case. This means we'll need to test new DSL support, and existing ILM support.

Here's a high-level overview of suggested tests, **which will need to be performed against both serverless and stateful instances**

- Test `[beat] setup --index-management` against a fresh ES instance, with all default settings
- Test `[beat] setup --index-management` against a fresh ES instance, with a custom policy file
- Test `[beat] setup --index-management` against a fresh ES instance, with both ILM and DSL enabled in the config, ensure beats fails
- Test `[beat] setup --index-management` against an existing ES instance with existing datastream config, all default settings, insure we don't overwrite
- Test `[beat] setup --index-management` against an existing ES instance with existing datastream config, with `overwrite` enabled, ensure we have overwritten data.
- Test `[beat] setup --index-management` with no `setup.*` flags set in the config, ensure defaults are properly set up.
- Run the beat against a new ES instance
- Run the beat against an existing ES instance that has existing setup data from an older beat version
- Test the setup process twice with `overwrite` enabled and a custom policy on the second setup. Ensure the policy is successfully updated.

### Tests to run over serverless only:
- set `setup.template.name` and `setup.dsl.policy_name` to different values, run `setup --index-management`. A warning message should be printed.
- set `setup.template.name` and `setup.dsl.policy_name` to the same value, run `setup --index-management`, then run again with a custom policy, and `setup.dsl.overwrite` set to true

### To create a custom policy for testing:

For DSL: create a file with `{"data_retention": "71d"}`, point to it with the `setup.dsl.policy_file` value

For ILM: ILM policies are a little more complex, see [here](https://www.elastic.co/guide/en/elasticsearch/reference/8.10/ilm-put-lifecycle.html) for an example, point to it with the `setup.ilm.policy_file` value

### Validating setup

To validate the `setup --index-management`, there are two ES endpoints to check: `_data_stream` points to the data stream, and contains a `lifecycle` section that should contain a valid lifecycle policy (see the above section for an example). The other endpoint is `_index_template`, which should contain a matching index template that links to the data stream.


## Known issues

- Due to the weird behavior of the `config` library in `elastic-agent-libs` (issue forthcoming), if a user has both `setup.ilm.*` and `setup.dsl.*` config values, one will need to be explicitly set to `disabled`, even if the other is explicitly set to `enabled`

## Config principles
Beats now has to care about both ILM and DSL config, as well as the upstream elasticsearch. Beats should behave based on these principles:

- If a user has enabled one lifecycle type but the upstream ES is the other type, return an error
- If a user has enabled both lifecycle types, return an error
- If the user has configured nothing, create the lifecycle policy based on the connected ES type
- During initial setup, create the lifecycle policy by injecting it directly into the template. To update the policy, perform a separate PUT call without modifying the template.

## Undefined behavior

Right now, there's some edge cases where I'm not sure about the correct behavior. 

- If the user defines a custom template name but _not_ a custom DSL policy name, then initial setup will work, but a second setup with `overwrite` enabled will fail, This is because the initial index management setup injects the lifecycle policy directly into the template policy. However, to update the DSL policy, we must make a separate REST call to an endpoint that includes the template name. If the user has set a custom template name but not a policy name, the code will revert to a default (and incorrect) template name. Should the user be expected to set `template.name` and `policy_name`? Should the code silently defer to `template.name`? Should the initial setup fail and tell the user to correct their config if one is set but not the other?
